mono/mini/mini-arm.c

```diff
@@ -634,6 +695,46 @@ emit_restore_lmf (MonoCompile *cfg, guint8 *code, gint32 lmf_offset)
 
 #endif /* #ifndef DISABLE_JIT */
 
+#ifndef MONO_CROSS_COMPILE
+static gboolean
+mono_arm_have_fast_tls (void)
+{
+#if (defined(HAVE_KW_THREAD) && defined(__linux__)) \
+	|| defined(TARGET_ANDROID)
+	guint32* kuser_get_tls = (void*)0xffff0fe0;
+	guint32 expected [] = {0xee1d0f70, 0xe12fff1e};
+
+	/* Expecting mrc + bx lr in the kuser_get_tls kernel helper */
+	return memcmp (kuser_get_tls, expected, 8) == 0;
+#elif defined(TARGET_IOS)
+	guint32 expected [] = {0x1f70ee1d, 0x0103f021, 0x0020f851, 0xbf004770};
+	/* Discard thumb bit */
+	guint32* pthread_getspecific_addr = (guint32*) ((guint32)pthread_getspecific & 0xfffffffe);
+	return memcmp ((void*)pthread_getspecific_addr, expected, 16) == 0;
```

Comment:
It should never assert. The fallback must be used when the fast path is not usable.
Asserting here would be a massive disaster as it would break all our existing customer apps.

As a rule of thumb only offline tools assert, the runtime should print a warning and fallback to something.

Having said that, the usage of the fallback must be controllable through MONO_DEBUG so we can test
it on existing devices.